src/utils/front-matter.js

```diff
@@ -5,38 +5,54 @@ const {
   builders: { hardline, concat, markAsRoot },
 } = require("../document");
 
+const yamlStart = "---";
+const yamlEnd = "...";
 const DELIMITER_MAP = {
-  "---": "yaml",
+  [yamlStart]: "yaml",
   "+++": "toml",
 };
 
 function parse(text) {
-  const delimiterRegex = Object.keys(DELIMITER_MAP).map(escape).join("|");
+  const delimiterStartRegex = Object.keys(DELIMITER_MAP).map(escape).join("|");
+  const delimiterEndRegex = Object.keys(DELIMITER_MAP)
+    .concat(yamlEnd)
+    .map(escape)
+    .join("|");
 
   const match = text.match(
     // trailing spaces after delimiters are allowed
     new RegExp(
-      `^(${delimiterRegex})([^\\n]*)\\n(?:|([\\s\\S]*?)\\n)\\1[^\\n\\S]*(\\n|$)`
+      `^(${delimiterStartRegex})([^\\n]*)\\n(?:|([\\s\\S]*?)\\n)(${delimiterEndRegex})[^\\n\\S]*(\\n|$)`
     )
   );
 
   if (match === null) {
     return { frontMatter: null, content: text };
   }
 
-  const [raw, delimiter, language, value = ""] = match;
-  let lang = DELIMITER_MAP[delimiter];
+  const [raw, delimiterStart, language, value = "", delimiterEnd] = match;
+  if (
+    delimiterStart !== delimiterEnd &&
+    (delimiterStart !== yamlStart || delimiterEnd !== yamlEnd)
+  ) {
+    return { frontMatter: null, content: text };
+  }
+  let lang = DELIMITER_MAP[delimiterStart];
   if (lang !== "toml" && language && language.trim()) {
     lang = language.trim();
   }
 
+  const frontMatter = {
+    type: "front-matter",
+    lang,
+    value,
+    raw: raw.replace(/\n$/, ""),
+  };
+  if (delimiterEnd === yamlEnd) {
+    frontMatter.yamlDocumentEndMarker = true;
```

Comment:
What do you think about setting `delimiterEnd`, not boolean?
```suggestion
    frontMatter.yamlEndDelimiter= delimiterEnd;
```
and use this in `print` function as:
```js
return markAsRoot(
  concat([yamlStart, hardline, doc, doc ? hardline : "", node.yamlEndDelimiter])
);
```

Reply:
done!

Reply:
Let's call it `frontMatter.openDelimiter`, and add `frontMatter.closeDelimiter` too

Reply:
done!